internal/datacoord/import_scheduler.go

```diff
@@ -0,0 +1,317 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+)
+
+const (
+	NullNodeID = -1
+)
+
+type ImportScheduler interface {
+	Start()
+	Close()
+}
+
+type importScheduler struct {
+	meta    *meta
+	cluster Cluster
+	alloc   allocator
+	imeta   ImportMeta
+
+	closeOnce sync.Once
+	closeChan chan struct{}
+}
+
+func NewImportScheduler(meta *meta,
+	cluster Cluster,
+	alloc allocator,
+	imeta ImportMeta,
+) ImportScheduler {
+	return &importScheduler{
+		meta:      meta,
+		cluster:   cluster,
+		alloc:     alloc,
+		imeta:     imeta,
+		closeChan: make(chan struct{}),
+	}
+}
+
+func (s *importScheduler) Start() {
+	log.Info("start import scheduler")
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.closeChan:
+			log.Info("import scheduler exited")
+			return
+		case <-ticker.C:
+			s.process()
+		}
+	}
+}
+
+func (s *importScheduler) Close() {
+	s.closeOnce.Do(func() {
+		close(s.closeChan)
+	})
+}
+
+func (s *importScheduler) process() {
+	all := s.imeta.GetBy()
+	tasksByJob := lo.GroupBy(all, func(t ImportTask) int64 {
+		return t.GetJobID()
+	})
+	jobs := lo.Keys(tasksByJob)
+	sort.Slice(jobs, func(i, j int) bool {
+		return jobs[i] < jobs[j]
+	})
+	for _, job := range jobs {
```

Comment:
During scheduling, the number of slots can be queried in advance, so there is no need to call grpc once for every task assigned